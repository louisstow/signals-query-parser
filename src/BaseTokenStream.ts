import type { InputStream } from "./InputStream";

type predicateFn = (c: string) => boolean;
type Token = {
  type: string;
  value: string;
  pos?: number;
};

class BaseTokenStream {
  istream: InputStream;
  current: Token | null;

  constructor(istream: InputStream) {
    this.istream = istream;
    this.current = null;
  }

  readNext(): Token | null {
    return null;
  }

  readWhile(predicate: predicateFn) {
    let s = "";
    while (!this.istream.eof() && predicate(this.istream.peek())) {
      s += this.istream.next();
    }
    return s;
  }

  readEscapedString(
    {
      keepEscape,
      expectClosingQuote,
    }: { keepEscape?: boolean; expectClosingQuote?: boolean } = {
      keepEscape: false,
      expectClosingQuote: true,
    }
  ) {
    let escaped = false;
    let closed = false;
    let s = "";

    while (!this.istream.eof()) {
      const c = this.istream.next();

      if (escaped) {
        if (keepEscape) s += "\\";
        s += c;
        escaped = false;
      } else if (c === "\\") {
        escaped = true;
      } else if (c === '"') {
        closed = true;
        break;
      } else {
        s += c;
      }
    }

    if (!closed && expectClosingQuote) {
      throw new Error("Unclosed string");
    }

    return s;
  }

  next() {
    const tok = this.current;
    this.current = null;
    return tok || this.readNext();
  }

  peek() {
    return this.current || (this.current = this.readNext());
  }

  eof() {
    return this.peek() === null;
  }
}

export { BaseTokenStream, Token };
