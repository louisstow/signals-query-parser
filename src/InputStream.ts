class InputStream {
  input: string;
  state: number[];
  pos: number;

  constructor(input: string) {
    this.state = [];
    this.pos = 0;
    this.input = input;
  }

  reset(input: string) {
    this.input = input;
    this.pos = 0;
  }

  next() {
    return this.input[this.pos++];
  }

  peek() {
    return this.input[this.pos];
  }

  eof() {
    return this.pos >= this.input.length;
  }

  // stack methods to go back in time if we find a dead-end
  save() {
    this.state.push(this.pos);
  }

  restore() {
    const pos = this.state.pop();
    if (pos !== undefined) {
      this.pos = pos;
    }
  }
}

export { InputStream };
