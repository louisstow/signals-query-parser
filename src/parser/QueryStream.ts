import type { TokenStream } from "./TokenStream";
import type { Token } from "../BaseTokenStream";

const isPipe = (t: Token | null) => t?.type === "pipe";
const isComma = (t: Token | null) => t?.type === "comma";
const isCloseParen = (t: Token | null) => t?.type === "closeParen";
const isOpenParen = (t: Token | null) => t?.type === "openParen";
const isSep = (t: Token | null) => isPipe(t) || isComma(t) || isCloseParen(t);
const isIdent = (t: Token | null): t is Token => t?.type === "ident";
const isText = (t: Token | null): t is Token =>
  t?.type === "string" || t?.type === "ident";
const isOperator = (t: Token | null): t is Token => t?.type === "operator";

const isIdentNoGuard = (t: Token | null) => t?.type === "ident";
const isTextNoGuard = (t: Token | null) =>
  t?.type === "string" || t?.type === "ident";

const getValidOp = (s: string) => {
  switch (s.toLowerCase()) {
    case "gt":
    case ">":
      return "gt";
    case "gte":
    case ">=":
      return "gte";
    case "lt":
    case "<":
      return "lt";
    case "lte":
    case "<=":
      return "lte";
    case "eq":
    case "=":
    case "==":
      return "eq";
  }

  return null;
};

const isValidMatchOp = (s: string) => {
  switch (s.toLowerCase()) {
    case "match":
    case "like":
    case "~":
      return true;
  }

  return false;
};

const isValidEqualOp = (s: string) => {
  switch (s.toLowerCase()) {
    case "=":
    case "eq":
    case "is":
    case "in":
      return true;
  }

  return false;
};

class ParserError extends Error {
  name = "ParserError";
  columnNumber: number;
  example: string;

  constructor(message: string, columnNumber: number, example: string) {
    super(message);
    this.columnNumber = columnNumber;
    this.example = example;
  }
}

class QueryStream {
  tstream: TokenStream;
  validFields: string[];

  constructor(tstream: TokenStream, validFields: string[]) {
    this.tstream = tstream;
    this.validFields = validFields;
  }

  createParserError(message: string, columnNumber?: number) {
    const input = this.tstream.istream.input;
    const pos =
      columnNumber ?? this.tstream.current?.pos ?? this.tstream.istream.pos;

    const start = Math.max(0, pos - 20);
    const end = Math.min(input.length, pos + 20);
    const offset = pos - start;

    const example = [input.slice(start, end), " ".repeat(offset) + "^"].join(
      "\n"
    );

    return new ParserError(message, pos, example);
  }

  readWhile(predicate: (t: Token) => boolean) {
    const values = [];
    let peek: Token | null = null;

    while (
      !this.tstream.eof() &&
      (peek = this.tstream.peek()) &&
      predicate(peek)
    ) {
      const t = this.tstream.next();
      if (t) {
        values.push(t);
      }
    }

    return values;
  }

  readSoftwareVersion() {
    let versions: Array<{ op: string; value: string }> = [];
    let peek: Token | null;

    while (
      !this.tstream.eof() &&
      (peek = this.tstream.peek()) &&
      !isComma(peek) &&
      !isCloseParen(peek)
    ) {
      const opToken = this.tstream.next();
      if (!opToken) {
        break;
      }

      const valueToken = this.tstream.next();

      if (!isTextNoGuard(valueToken)) {
        throw this.createParserError(`Expected version`, valueToken?.pos);
      }

      if (isText(valueToken)) {
        const op = getValidOp(opToken.value);
        if (!op) {
          throw this.createParserError(
            `Invalid version operator '${opToken.value}'`,
            opToken.pos
          );
        }

        versions.push({
          op,
          value: valueToken.value,
        });
      }
    }

    return versions;
  }

  readSoftwareInValue() {
    const softwareToken = this.tstream.next();
    const name = isText(softwareToken) ? softwareToken.value : null;

    if (!name) {
      throw this.createParserError(
        `Expected software name`,
        softwareToken?.pos
      );
    }

    const peek = this.tstream.peek();
    const version = peek && !isSep(peek) ? this.readSoftwareVersion() : [];

    return {
      name,
      version,
    };
  }

  readSoftwareIn() {
    const software = [];
    let peek: Token | null;

    while (
      !this.tstream.eof() &&
      (peek = this.tstream.peek()) &&
      !isCloseParen(peek) &&
      !isPipe(peek)
    ) {
      this.readWhile(isComma);

      const s = this.readSoftwareInValue();
      if (s) {
        software.push(s);
      }
    }

    return software;
  }

  readSoftwareQuery() {
    const fn = this.tstream.next();

    if (!isText(fn) && !isOperator(fn)) {
      throw this.createParserError(`Unexpected operator`);
    }

    const op = fn.value.toLowerCase();
    if (!isValidEqualOp(op) && !isValidMatchOp(op)) {
      throw this.createParserError(`Unexpected operator '${fn.value}'`, fn.pos);
    }

    const fuzzy = isValidMatchOp(op);

    let hasOpenParen = false;

    if (isOpenParen(this.tstream.peek())) {
      this.tstream.next();
      hasOpenParen = true;
    }

    if (isText(this.tstream.peek())) {
      // check software query by starting with @
      const token = this.tstream.peek();

      if (isIdent(token) && token.value.startsWith("@")) {
        this.tstream.next();

        return {
          type: "software",
          value: {
            software: [],
            softwareList: token.value,
            fuzzy,
          },
        };
      }

      const software = this.readSoftwareIn();
      if (isCloseParen(this.tstream.peek())) {
        this.tstream.next();
      } else if (hasOpenParen) {
        throw this.createParserError(`Expected closing parenthesis`);
      }

      return {
        type: "software",
        value: {
          software,
          fuzzy,
        },
      };
    }

    throw this.createParserError(`Expected software query`);
  }

  readValue(field: string, type: string) {
    const valueToken = this.tstream.next();

    if (!isTextNoGuard(valueToken)) {
      throw this.createParserError(
        `Unexpected value for field '${field}'`,
        valueToken?.pos
      );
    }

    if (isText(valueToken)) {
      return {
        type,
        value: {
          field,
          query: valueToken.value,
        },
      };
    }
  }

  readIn(field: string, negated?: boolean) {
    const query: string[] = [];
    let peek: Token | null;
    let hasOpenParen = false;

    while (
      !this.tstream.eof() &&
      (peek = this.tstream.peek()) &&
      !isCloseParen(peek) &&
      !isPipe(peek)
    ) {
      this.readWhile(isComma);

      const t = this.tstream.next();

      if (isOpenParen(t)) {
        hasOpenParen = true;
      }

      if (isText(t)) {
        query.push(t.value);
      }
    }

    if (isCloseParen(this.tstream.peek())) {
      this.tstream.next();
    } else if (hasOpenParen) {
      throw this.createParserError(`Expected closing parenthesis`);
    }

    return {
      type: "in",
      value: {
        field,
        query,
        negated: negated ? true : undefined,
      },
    };
  }

  readRange(field: string, op: string) {
    const query: { [k: string]: number } = {};
    let peek: Token | null;

    const firstValue = this.tstream.next();

    if (!isIdentNoGuard(firstValue)) {
      throw this.createParserError(
        `Expected a number for field '${field}'`,
        firstValue?.pos
      );
    }

    if (isIdent(firstValue)) {
      query[op] = Number(firstValue.value);

      if (isNaN(query[op])) {
        throw this.createParserError(
          `Not a number '${firstValue.value}'`,
          firstValue.pos
        );
      }
    }

    while (
      !this.tstream.eof() &&
      (peek = this.tstream.peek()) &&
      !isPipe(peek)
    ) {
      const t = this.tstream.next();
      const validOp = (isText(t) || isOperator(t)) && getValidOp(t.value);

      if (validOp) {
        const v = this.tstream.next();

        if (!isIdentNoGuard(v)) {
          throw this.createParserError(
            `Expected a number for field '${field}'`,
            v?.pos
          );
        }

        if (isIdent(v)) {
          query[validOp] = Number(v.value);

          if (isNaN(query[validOp])) {
            throw this.createParserError(
              `Not a number '${v.value}' for field '${field}'`,
              v.pos
            );
          }
        }
      }
    }

    return {
      type: "range",
      value: {
        field,
        query,
      },
    };
  }

  readFilter() {
    const filter = this.tstream.next();
    if (!filter) {
      return null;
    }

    return {
      type: "filter",
      value: {
        filter: filter.value,
      },
    };
  }

  readGroup() {
    const group = this.tstream.next();
    if (!group) {
      return null;
    }

    return {
      type: "group",
      value: {
        field: group.value,
      },
    };
  }

  readNegated() {
    const token = this.tstream.peek();
    if (isIdent(token) && token.value == "not") {
      this.tstream.next();
      return true;
    }

    return false;
  }

  readQuery() {
    const field = this.tstream.next();

    if (!field) {
      return null;
    }

    if (field.type === "ident" && field.value === "filter") {
      return this.readFilter();
    }

    if (field.type === "ident" && field.value === "group") {
      return this.readGroup();
    }

    if (field.type === "ident" && field.value === "software") {
      return this.readSoftwareQuery();
    }

    if (!this.validFields.includes(field.value)) {
      throw this.createParserError(`Invalid field '${field.value}'`, field.pos);
    }

    const negated = this.readNegated();
    const fn = this.tstream.next();

    if (isIdent(fn) || isOperator(fn)) {
      if (fn.value.toLowerCase() === "in") {
        return this.readIn(field.value, negated);
      }

      if (isValidMatchOp(fn.value)) {
        return this.readValue(field.value, "match");
      }

      if (isValidEqualOp(fn.value)) {
        return this.readValue(field.value, "exact");
      }

      const op = getValidOp(fn.value);
      if (op !== null) {
        return this.readRange(field.value, op);
      } else {
        throw this.createParserError(`Invalid operator '${fn.value}'`, fn.pos);
      }
    }
  }

  parseToken() {
    this.readWhile(isPipe);

    if (this.tstream.eof()) {
      return null;
    }

    const token = this.tstream.peek();
    if (!token) {
      return null;
    }

    if (token.type === "ident") {
      return this.readQuery();
    }

    if (token.type === "string") {
      this.tstream.next();
      return {
        type: "match",
        value: {
          field: "aggregate",
          query: token.value,
        },
      };
    }

    throw new Error(`Unexpected token: ${token.type} ${token.value}`);
  }

  parse() {
    const q: any[] = [];
    while (!this.tstream.eof()) {
      try {
        const queryItem = this.parseToken();

        if (queryItem) {
          q.push(queryItem);
        }
      } catch (err) {
        if (err instanceof ParserError) {
          throw err;
        } else {
          const error = err as Error;
          throw this.createParserError(error.message);
        }
      }
    }

    return q;
  }
}

export { QueryStream, ParserError };
