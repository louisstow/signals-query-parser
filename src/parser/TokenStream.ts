import { BaseTokenStream } from "../BaseTokenStream";
import type { InputStream } from "../InputStream";

const PIPE = "|";
const QUOTE = '"';
const OPEN_PAREN = "(";
const CLOSE_PAREN = ")";
const COMMA = ",";

const isNumeric = (c: string) => /^[0-9\.\-]+$/.test(c);
const isSep = (c: string) =>
  c === PIPE || c === OPEN_PAREN || c === CLOSE_PAREN || c === COMMA;
const isText = (c: string) => !isNumeric(c);
const isWhitespace = (c: string) =>
  c === " " || c === "\t" || c === "\n" || c === "\r";
const isIdent = (c: string) => !isWhitespace(c) && !isSep(c);

class TokenStream extends BaseTokenStream {
  constructor(input: InputStream) {
    super(input);
  }

  readString(expectClosingQuote = true) {
    return {
      pos: this.istream.pos,
      type: "string",
      value: this.readEscapedString({ expectClosingQuote }),
    };
  }

  readNumber() {
    return {
      pos: this.istream.pos,
      type: "number",
      value: this.readWhile(isNumeric),
    };
  }

  readIdent() {
    return {
      pos: this.istream.pos,
      type: "ident",
      value: this.readWhile(isIdent),
    };
  }

  readNext() {
    this.readWhile(isWhitespace);

    if (this.istream.eof()) {
      return null;
    }

    const c = this.istream.peek();

    if (c === PIPE) {
      this.istream.next();
      return {
        type: "pipe",
        value: c,
        pos: this.istream.pos - 1,
      };
    }

    if (c === OPEN_PAREN) {
      this.istream.next();
      return {
        type: "openParen",
        value: c,
        pos: this.istream.pos - 1,
      };
    }

    if (c === CLOSE_PAREN) {
      this.istream.next();
      return {
        type: "closeParen",
        value: c,
        pos: this.istream.pos - 1,
      };
    }

    if (c === COMMA) {
      this.istream.next();
      return {
        type: "comma",
        value: c,
        pos: this.istream.pos - 1,
      };
    }

    if (c === QUOTE) {
      this.istream.next();
      return this.readString(true);
    }

    // if (isNumeric(c)) {
    //   return this.readNumber();
    // }

    if (isIdent(c)) {
      return this.readIdent();
    }

    throw new Error(
      `Unable to parse query character: ${c} at ${this.istream.pos}`
    );
  }
}

/**
 * RangeQuery: severity < 2 > 3
 * InQuery: collectors in ("12")
 * MatchQuery: description match "my string"
 * ExactQuery: id = "CVE-2020-123"
 * SoftwareQuery: software in ("Adobe Acrobat" gt 1.3)
 * CanonicalQuery: software in @software_list
 * CanonicalQuery: software = @software_list
 * SoftwareQuery: software = "Adobe Acrobat"
 */

export { TokenStream };
