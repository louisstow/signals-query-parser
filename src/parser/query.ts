import { InputStream } from "../InputStream";
import { TokenStream } from "./TokenStream";
import { ParserError, QueryStream } from "./QueryStream";
import type { Query } from "../query.types";

const parseQuery = (s: string, validFields: string[]) => {
  const inputStream = new InputStream(s);
  const tokenStream = new TokenStream(inputStream);
  const queryStream = new QueryStream(tokenStream, validFields);

  return queryStream.parse() as Query[];
};

export { parseQuery, ParserError };
