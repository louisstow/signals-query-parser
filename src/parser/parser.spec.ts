import { InputStream } from "../InputStream";
import { TokenStream } from "./TokenStream";
import { ParserError, QueryStream } from "./QueryStream";

import { parseQuery as originalParseQuery } from "./query";
import { Token } from "../BaseTokenStream";

const parseQuery = (query: string) => {
  return originalParseQuery(query, [
    "software",
    "softwareList",
    "severity",
    "description",
    "collectors",
    "id",
    "tags",
  ]);
};

describe("TokenStream", () => {
  const parse = (s: string) => {
    const inputStream = new InputStream(s);
    const tokenStream = new TokenStream(inputStream);
    const tokens: Token[] = [];

    while (!tokenStream.eof()) {
      const t = tokenStream.next();

      if (t) {
        tokens.push(t);
      }
    }

    return tokens;
  };

  test("SoftwareQuery", () => {
    expect(parse(`software in ("test" gt 1.2)`)).toEqual([
      { pos: 0, type: "ident", value: "software" },
      { pos: 9, type: "ident", value: "in" },
      { pos: 12, type: "openParen", value: "(" },
      { pos: 14, type: "string", value: "test" },
      { pos: 20, type: "ident", value: "gt" },
      { pos: 23, type: "ident", value: "1.2" },
      { pos: 26, type: "closeParen", value: ")" },
    ]);

    expect(parse(`software in ("test" =1.2>3)`)).toEqual([
      { pos: 0, type: "ident", value: "software" },
      { pos: 9, type: "ident", value: "in" },
      { pos: 12, type: "openParen", value: "(" },
      { pos: 14, type: "string", value: "test" },
      { pos: 20, type: "operator", value: "=" },
      { pos: 21, type: "ident", value: "1.2" },
      { pos: 24, type: "operator", value: ">" },
      { pos: 25, type: "ident", value: "3" },
      { pos: 26, type: "closeParen", value: ")" },
    ]);

    expect(parse(`software in (test = 1.2, another > 5 < 10)`)).toStrictEqual([
      { pos: 0, type: "ident", value: "software" },
      { pos: 9, type: "ident", value: "in" },
      { pos: 12, type: "openParen", value: "(" },
      { pos: 13, type: "ident", value: "test" },
      { pos: 18, type: "operator", value: "=" },
      { pos: 20, type: "ident", value: "1.2" },
      { pos: 23, type: "comma", value: "," },
      { pos: 25, type: "ident", value: "another" },
      { pos: 33, type: "operator", value: ">" },
      { pos: 35, type: "ident", value: "5" },
      { pos: 37, type: "operator", value: "<" },
      { pos: 39, type: "ident", value: "10" },
      { pos: 41, type: "closeParen", value: ")" },
    ]);
  });
});

test("QueryStream", () => {
  expect(
    parseQuery(`software in (test = 1.2, another > 5 < 10)`)
  ).toMatchSnapshot();

  expect(
    parseQuery(`software in (test =1.2, another >5<10)`)
  ).toMatchSnapshot();

  expect(parseQuery(`software in (test, another)`)).toMatchSnapshot();

  expect(
    parseQuery(`software = test = 1.2, "another" > 5 < 10`)
  ).toMatchSnapshot();

  expect(parseQuery(`description match "my query"`)).toMatchSnapshot();

  expect(
    parseQuery(`collectors in ("a:thing", "c:another")`)
  ).toMatchSnapshot();

  expect(parseQuery(`severity > 2 < 5`)).toMatchSnapshot();

  expect(parseQuery(`id = CVE-2020-123`)).toMatchSnapshot();

  expect(
    parseQuery(`id = CVE-2020-123 | severity > 4 | description like "Apple"`)
  ).toMatchSnapshot();

  expect(parseQuery(`"CVE-2020-123"`)).toMatchSnapshot();

  expect(parseQuery(`software in @software_list`)).toMatchSnapshot();

  expect(parseQuery(`software = @software_list`)).toMatchSnapshot();

  expect(
    parseQuery(`software = @software_list | filter THING`)
  ).toMatchSnapshot();

  expect(parseQuery(`tags in ("a", "b")`)).toMatchSnapshot();

  expect(parseQuery(`tags not in ("a", "b")`)).toMatchSnapshot();
});

test("Grouping", () => {
  expect(
    parseQuery(`tags in ("vendor/microsoft") | group tags`)
  ).toMatchSnapshot();
});

test("Error handling", () => {
  try {
    parseQuery(`thing`);
  } catch (err) {
    const e = err as ParserError;
    expect(e.message).toMatch(/Invalid field/);
    expect(e.columnNumber).toBe(0);
    expect(e.example).toMatchSnapshot();
  }

  try {
    parseQuery(`severity >`);
  } catch (err) {
    const e = err as ParserError;
    expect(e.message).toMatch(/Expected a number/);
    expect(e.columnNumber).toBe(10);
    expect(e.example).toMatchSnapshot();
  }

  try {
    parseQuery(`severity > "test"`);
  } catch (err) {
    const e = err as ParserError;
    expect(e.message).toMatch(/Expected a number/);
    expect(e.columnNumber).toBe(12);
    expect(e.example).toMatchSnapshot();
  }

  try {
    parseQuery(`severity > x`);
  } catch (err) {
    const e = err as ParserError;
    expect(e.message).toMatch(/Not a number/);
    expect(e.columnNumber).toBe(11);
    expect(e.example).toMatchSnapshot();
  }

  try {
    parseQuery(`description ~`);
  } catch (err) {
    const e = err as ParserError;
    expect(e.message).toMatch(/Unexpected value/);
    expect(e.columnNumber).toBe(13);
    expect(e.example).toMatchSnapshot();
  }

  try {
    parseQuery(`description # "test"`);
  } catch (err) {
    const e = err as ParserError;
    expect(e.message).toMatch(/Invalid operator/);
    expect(e.columnNumber).toBe(12);
    expect(e.example).toMatchSnapshot();
  }

  try {
    parseQuery(`software in x = , y > 2`);
  } catch (err) {
    const e = err as ParserError;
    expect(e.message).toMatch(/Expected version/);
    expect(e.columnNumber).toBe(16);
    expect(e.example).toMatchSnapshot();
  }

  try {
    parseQuery(`software in x % 2, y > 2`);
  } catch (err) {
    const e = err as ParserError;
    expect(e.message).toMatch(/Invalid version operator/);
    expect(e.columnNumber).toBe(14);
    expect(e.example).toMatchSnapshot();
  }

  try {
    parseQuery(`software in |`);
  } catch (err) {
    const e = err as ParserError;
    expect(e.message).toMatch(/Expected software query/);
    expect(e.columnNumber).toBe(12);
    expect(e.example).toMatchSnapshot();
  }

  try {
    parseQuery(
      `description in "this is a very long description" | software in |`
    );
  } catch (err) {
    const e = err as ParserError;
    expect(e.message).toMatch(/Expected software query/);
    expect(e.columnNumber).toBe(63);
    expect(e.example).toMatchSnapshot();
  }

  try {
    parseQuery(`collectors in (a, ")`);
  } catch (err) {
    const e = err as ParserError;
    expect(e.message).toMatch(/Unclosed string/);
    expect(e.columnNumber).toBe(20);
    expect(e.example).toMatchSnapshot();
  }

  try {
    parseQuery(`collectors in (a, b`);
  } catch (err) {
    const e = err as ParserError;
    expect(e.message).toMatch(/Expected closing parenthesis/);
    expect(e.columnNumber).toBe(19);
    expect(e.example).toMatchSnapshot();
  }

  try {
    parseQuery(`software in (x = 1, y > 2`);
  } catch (err) {
    const e = err as ParserError;
    expect(e.message).toMatch(/Expected closing parenthesis/);
    expect(e.columnNumber).toBe(25);
    expect(e.example).toMatchSnapshot();
  }

  expect.hasAssertions();
});

describe("Misc tests", () => {
  // expect(parseQuery(`collectors in (a, ")`)).toEqual([
  //   {
  //     type: "in",
  //     value: {
  //       field: "collectors",
  //       query: ["a"],
  //     },
  //   },
  // ]);
});
