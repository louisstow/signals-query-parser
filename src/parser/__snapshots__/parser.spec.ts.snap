// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error handling 1`] = `
"thing
^"
`;

exports[`Error handling 2`] = `
"severity >
          ^"
`;

exports[`Error handling 3`] = `
"severity > \\"test\\"
            ^"
`;

exports[`Error handling 4`] = `
"severity > x
           ^"
`;

exports[`Error handling 5`] = `
"description ~
             ^"
`;

exports[`Error handling 6`] = `
"description # \\"test\\"
            ^"
`;

exports[`Error handling 7`] = `
"software in x = , y > 2
                ^"
`;

exports[`Error handling 8`] = `
"software in x % 2, y > 2
              ^"
`;

exports[`Error handling 9`] = `
"software in |
            ^"
`;

exports[`Error handling 10`] = `
"tion\\" | software in |
                    ^"
`;

exports[`QueryStream 1`] = `
Array [
  Object {
    "type": "software",
    "value": Object {
      "fuzzy": false,
      "software": Array [
        Object {
          "name": "test",
          "version": Array [
            Object {
              "op": "eq",
              "value": "1.2",
            },
          ],
        },
        Object {
          "name": "another",
          "version": Array [
            Object {
              "op": "gt",
              "value": "5",
            },
            Object {
              "op": "lt",
              "value": "10",
            },
          ],
        },
      ],
    },
  },
]
`;

exports[`QueryStream 2`] = `
Array [
  Object {
    "type": "software",
    "value": Object {
      "fuzzy": false,
      "software": Array [
        Object {
          "name": "test",
          "version": Array [],
        },
        Object {
          "name": "another",
          "version": Array [],
        },
      ],
    },
  },
]
`;

exports[`QueryStream 3`] = `
Array [
  Object {
    "type": "software",
    "value": Object {
      "fuzzy": false,
      "software": Array [
        Object {
          "name": "test",
          "version": Array [
            Object {
              "op": "eq",
              "value": "1.2",
            },
          ],
        },
        Object {
          "name": "another",
          "version": Array [
            Object {
              "op": "gt",
              "value": "5",
            },
            Object {
              "op": "lt",
              "value": "10",
            },
          ],
        },
      ],
    },
  },
]
`;

exports[`QueryStream 4`] = `
Array [
  Object {
    "type": "match",
    "value": Object {
      "field": "description",
      "query": "my query",
    },
  },
]
`;

exports[`QueryStream 5`] = `
Array [
  Object {
    "type": "in",
    "value": Object {
      "field": "collectors",
      "query": Array [
        "a:thing",
        "c:another",
      ],
    },
  },
]
`;

exports[`QueryStream 6`] = `
Array [
  Object {
    "type": "range",
    "value": Object {
      "field": "severity",
      "query": Object {
        "gt": 2,
        "lt": 5,
      },
    },
  },
]
`;

exports[`QueryStream 7`] = `
Array [
  Object {
    "type": "exact",
    "value": Object {
      "field": "id",
      "query": "CVE-2020-123",
    },
  },
]
`;

exports[`QueryStream 8`] = `
Array [
  Object {
    "type": "exact",
    "value": Object {
      "field": "id",
      "query": "CVE-2020-123",
    },
  },
  Object {
    "type": "range",
    "value": Object {
      "field": "severity",
      "query": Object {
        "gt": 4,
      },
    },
  },
  Object {
    "type": "match",
    "value": Object {
      "field": "description",
      "query": "Apple",
    },
  },
]
`;

exports[`QueryStream 9`] = `
Array [
  Object {
    "type": "match",
    "value": Object {
      "field": "aggregate",
      "query": "CVE-2020-123",
    },
  },
]
`;

exports[`QueryStream 10`] = `
Array [
  Object {
    "type": "software",
    "value": Object {
      "fuzzy": false,
      "software": Array [],
      "softwareList": "@software_list",
    },
  },
]
`;

exports[`QueryStream 11`] = `
Array [
  Object {
    "type": "software",
    "value": Object {
      "fuzzy": false,
      "software": Array [],
      "softwareList": "@software_list",
    },
  },
]
`;

exports[`QueryStream 12`] = `
Array [
  Object {
    "type": "software",
    "value": Object {
      "fuzzy": false,
      "software": Array [],
      "softwareList": "@software_list",
    },
  },
  Object {
    "type": "filter",
    "value": Object {
      "filter": "THING",
    },
  },
]
`;
